// @fpjs-incubator/broyster/dist/browserstack_capabilities_factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserStackCapabilitiesFactory = void 0;
class BrowserStackCapabilitiesFactory {
    constructor(_credentials, _local = true) {
        this._credentials = _credentials;
        this._local = _local;
    }
    create(browserName, buildName, sessionName, projectName, deviceName, os, idleTimeout, osVersion, browserVersion, localIdentifier) {
        return {
            'bstack:options': {
                os: os,
                osVersion: osVersion,
                deviceName: deviceName,
                projectName: projectName,
                sessionName: sessionName,
                buildName: buildName,
                local: this._local,
                userName: this._credentials.username,
                accessKey: this._credentials.accessKey,
                idleTimeout: idleTimeout,
                localIdentifier: localIdentifier,
                networkLogs: true,
                networkLogsOptions: {
                    captureContent: 'true',
                },
                consoleLogs: 'verbose',
            },
            browserName: browserName === null || browserName === void 0 ? void 0 : browserName.toLowerCase(),
            browserVersion: browserVersion || 'latest',
            acceptInsecureCerts: true,
        };
    }
}
exports.BrowserStackCapabilitiesFactory = BrowserStackCapabilitiesFactory;
