// @fpjs-incubator/broyster/dist/karma_plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const browser_map_1 = require("./browser_map");
const browserstack_local_manager_1 = require("./browserstack_local_manager");
const browserstack_session_factory_1 = require("./browserstack_session_factory");
const browserstack_sessions_manager_1 = require("./browserstack_sessions_manager");
const browserstack_browsers_1 = require("./browserstack_browsers");
const browserstack_helpers_1 = require("./browserstack_helpers");
const launcher_1 = require("./launcher");
const browserstack_reporter_1 = require("./browserstack_reporter");
const karmaPlugin = {
    'launcher:BrowserStack': ['type', launcher_1.BrowserStackLauncher],
    'reporter:BrowserStack': ['type', browserstack_reporter_1.BrowserStackReporter],
    browserStackSessionFactory: ['factory', browserstack_session_factory_1.makeBrowserStackSessionFactory],
    browserStackLocalManager: ['factory', browserstack_local_manager_1.makeBrowserStackLocalManagerFactory],
    browserStackSessionsManager: ['factory', browserstack_sessions_manager_1.makeBrowserStackSessionsManager],
    browserStackCredentials: ['factory', browserstack_helpers_1.getBrowserStackCredentials],
    browserStackBrowsers: ['factory', browserstack_browsers_1.makeBrowserStackBrowsers],
    browserMap: ['factory', browser_map_1.makeBrowserMapFactory],
    localIdentifier: ['factory', browserstack_local_manager_1.makeLocalIdentifier],
};
exports.default = karmaPlugin;
